src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -69,15 +81,35 @@ internal sealed class VisualStudioProject
         /// </summary>
         private readonly FileChangeWatcher.IContext _fileReferenceChangeContext;
 
+        /// <summary>
+        /// the map of <see cref="DocumentId"/> to <see cref="IDynamicFileInfoProvider"/> whose <see cref="DynamicFileInfo"/> got added into <see cref="Workspace"/>
+        /// </summary>
+        private readonly Dictionary<DocumentId, IDynamicFileInfoProvider> _documentIdToDynamicFileInfoProvider = new Dictionary<DocumentId, IDynamicFileInfoProvider>();
+
+        /// <summary>
+        /// track whether we have been subscribed to <see cref="IDynamicFileInfoProvider.Updated"/> event
```

Comment:
I am doing that to lazily subscribe to the change event. but we could pre-emptively subscribed to all providers if we want to.

but, let's do that in follow up PR. rather than this one.

Reply:
issue - https://github.com/dotnet/roslyn/issues/30891